### PR TITLE
gsl 2.2

### PR DESCRIPTION
**Pull request by ilovezfs**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by ilovezfs** [comment]
@dpo @sjackman @DomT4 :ship:?

**Comment by dpo** [comment]
What's the worry?

**Comment by ilovezfs** [comment]
```
dieharder                                          homebrew/science/gromacs                         
flowgrind                                          homebrew/science/hotspot                         
gnuradio                                           homebrew/science/libsequence                     
orpie                                              homebrew/science/lobstr                          
oscats                                             homebrew/science/mash                            
homebrew/gui/inkscape                              homebrew/science/mathgl                          
homebrew/science/abinit                            homebrew/science/methpipe                        
homebrew/science/analysis                          homebrew/science/moose                           
homebrew/science/andi                              homebrew/science/nco                             
homebrew/science/apophenia                         homebrew/science/nest                            
homebrew/science/art                               homebrew/science/nip2                            
homebrew/science/astrometry-net                    homebrew/science/pspp                            
homebrew/science/atomic-pseudopotential-engine     homebrew/science/qgis                            
homebrew/science/dynare                            homebrew/science/root                            
homebrew/science/ea-utils                          homebrew/science/root6                           
homebrew/science/enblend-enfuse                    homebrew/science/visp                            
homebrew/science/fgsl                              homebrew/science/xmi-msim                        
homebrew/science/fwdpp                             homebrew/x11/pyxplot                             
homebrew/science/getdp                             homebrew/x11/yarp                                
homebrew/science/gnudatalanguage                 
```
All will need to be rebuilt.

And the original alarm in
```
-# Please don't update me to the 2.x branch yet until issues discussed in
-# https://github.com/Homebrew/homebrew/issues/45812 are resolved.
-# If you want 2.x now, file a PR in homebrew/versions. Thanks!
```
https://github.com/Homebrew/homebrew/issues/45812

**Comment by sjackman** [comment]
I'm fine in principle, but know nothing about the reasoning for the above warning, which was written by Dominyk @DomT4. See https://github.com/Homebrew/homebrew-core/commit/9ee284fac333e2e96e1aa21419f20be4915a719b

**Comment by ilovezfs** [comment]
@sjackman I think enough time has gone by since the original 2 series came out that things have basically caught up.

I've so far locally been able to build these from science against gsl 2
```
abinit
analysis
andi
apophenia
art
astrometry-net
atomic-pseudopotential-engine
ea-utils
enblend-enfuse
fgsl
fwdpp
getdp
gnudatalanguage
hotspot
lobstr
mash
mathgl
methpipe
nco
nest
nip2
pspp
root
root6
visp
yarp
```

dynare has an unrelated case-sensitivity bug so I'm not sure.

gromacs gave me a
```
CMake Warning:
  Manually-specified variables were not used by the project:

    GMX_GSL
```
and didn't link against gsl but built, so not sure what the deal is there.

libsequence doesn't actually seem to use gsl.

pyqxplot dies (probably some unrelated library with an issue):
```
clang -Wall -g  obj/helpers/pyxplot_fitshelper.o obj/stringTools/asciidouble.o         -lcfitsio -lfftw3 `xml2-config --libs` `gsl-config --libs` -ltermcap -lreadline -lz -lpng  -lm -o bin/pyxplot_fitshelper
Undefined symbols for architecture x86_64:
  "_history_list", referenced from:
      _ppl_directive_history in core.o
      _ppl_directive_save in core.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [bin/pyxplot] Error 1
make: *** Waiting for unfinished jobs....
```

qgis is unhappy for an unrelated reason it seems:
```
In file included from /tmp/qgis-20160830-72146-1siuyb1/qgis-1.8.0/src/app/gps/qwtpolar-1.0/qwt_polar_itemdict.h:15:
/tmp/qgis-20160830-72146-1siuyb1/qgis-1.8.0/src/app/gps/qwtpolar-1.0/qwt_polar_item.h:14:10: fatal error: 
      'qwt_legend_itemmanager.h' file not found
#include <qwt_legend_itemmanager.h>
         ^
1 error generated.
make[2]: *** [src/app/CMakeFiles/QGIS.dir/qgisapp.cpp.o] Error 1
make[1]: *** [src/app/CMakeFiles/QGIS.dir/all] Error 2
make: *** [all] Error 2
```

`xmi-msim` is broken with gsl 2 but its HEAD at https://github.com/tschoonj/xmimsim.git is fine, so a new release from upstream would be good there.

So nothing too scary as far as I can tell.

**Comment by ilovezfs** [comment]
@tschoonj any thoughts on gsl 2 now that some time has passed?

**Comment by tschoonj** [comment]
Probably best to update at this point but some older formulas may need to be compiled against the old 1.x, which should probably be added to Homebrew-versions.

**Comment by ilovezfs** [comment]
@tschoonj versions has https://github.com/Homebrew/homebrew-versions/blob/master/gsl114.rb

Would you also add gsl116 or unnecessary?

**Comment by tschoonj** [comment]
I would replace 114 with 116 as the latter is fully backwards compatible.

**Comment by ilovezfs** [comment]
@tschoonj perfect that makes a lot of sense. I'm testing everything in science that uses gsl right now (https://bot.brew.sh/job/Homebrew%20Testing/1053/) to see what's already broken BEFORE this is merged, and then I'll ship this PR.

**Comment by ilovezfs** [comment]
@tschoonj by the way I think you're upstream for xmi-msim, right? It looks like latest actual tag isn't working with gsl 2 but HEAD is? If that's correct, any chance of getting a new release? :innocent:

**Comment by tschoonj** [comment]
Not ready for a new release yet, but I will add a patch for it later today.

Thanks for reminding me 😄

**Comment by ilovezfs** [comment]
@tschoonj yeah a patch for current stable is more-than-good-enough-for-me!

**Comment by tschoonj** [comment]
Cherrypicking commits for patching is not going to work unfortunately as it breaks a couple of things.

I suggest I update the xmi-msim formula with a `:DATA` patch, after you update the `gsl` and `fgsl` formulas (the latter is at 1.1.0, but I haven't update the formula since it depends on gsl 2.x).

Would that work for you?

**Comment by ilovezfs** [comment]
@tschoonj That would be fine (although it should be patch in https://github.com/Homebrew/formula-patches instead of :DATA), but wouldn't it be easier for everyone if you created a maintenance tag containing the patch, and we build from the automatically available tarball?

**Comment by tschoonj** [comment]
I haven't been very active with Homebrew for a couple of months now and I am embarrassed to admit that I don't fully understand what you are saying:

* Has the usage of `:DATA` blocks been deprecated/removed from Formulas? How does this formula-patches repo work?
* What's a maintenance tag? Do I create an issue for the `xmi-msim` formula and add a tag named like that? 😕 
* The main problem for me is that I cannot really write and test the patch until the `gsl` and `fgsl` formulas have both been updated to their latest versions and have been made available via `brew update`

Sorry about the confusion!

**Comment by ilovezfs** [comment]
> and I am embarrassed

don't be

> Has the usage of :DATA blocks been deprecated/removed from Formulas?

yeah we get rid of them where possible now as we go along but they still work if present

> How does this formula-patches repo work?

You open a pull request to that repository adding a new file foo.patch to a new subdirectory (if your formula doesn't already have any patches in the repo) with the name of your formula).

`brew install hub; git clone Homebrew/formula-patches; cd formula-patches; hub fork; git checkout -b xmi-msim-gsl2; mkdir xmi-msim; cp ~/foo.patch xmi-msim/gsl2.patch; git add xmi-msim/gsl2.patch ; git commit -am "xmi-msim: patch for gsl2"; git push -u tschoonj xmi-msim-gsl2; hub pull-request --browse`

Then we merge that PR. Which gives you a URL like `url "https://raw.githubusercontent.com/Homebrew/formula-patches/e8fe6567/qt5/restore-pc-files.patch"` and you use that in a `patch do` block. Note the use of the raw.githubusercontent.com link since you want the raw contents of the patch file, not the diff of adding the patch file to the formula-patches repo. The `patch do` should also contain the `sha256 "abcd..."` of the patch file. Note that you can test and open your PR using the URL from your fork before we actually merge the patch into the formula-patches repo. The checksum won't even change when you later update the URL to point to the formula-patches repo since the url is a link to the actual patch file which is the same regardless of which repository it's coming from.

>What's a maintenance tag? Do I create an issue for the xmi-msim formula and add a tag named like that?

Creating any git tag (standard annotated tag or gpg signed tag) and pushing it to github will automatically make a tarball containing the contents of the repository at that tag. That's what most of the tarballs in homebrew are, just automatically available archives generated by github from the tags in the repo. It's a separate procedure if you actually want to create a tag that has its own release tarball (e.g., containing generated code such as a configure file post autogen) to avoid autotools dependencies or exclude certain files, etc.

is an example of the simple case where there are only automatically generated archives based on the corresponding tag. For instance

https://github.com/tschoonj/xmimsim/releases
https://github.com/tschoonj/xmimsim/tags

commit:
https://github.com/tschoonj/xmimsim/commit/282b0ec433bca3523c6e4963fdbdafe2cc6a54a4

(unsigned) annotated tag:
https://github.com/tschoonj/xmimsim/releases/tag/XMI-MSIM-5.0

automatically generated release tarball:
https://github.com/tschoonj/xmimsim/archive/XMI-MSIM-5.0.tar.gz

So you can
```
git checkout -b XMI-MSIM-5.0.1-gsl2-branch XMI-MSIM-5.0
...make the changes...
git commit -am "XMI-MSIM-5.0.1: adapt 5.0 for gsl2"
git push -u origin XMI-MSIM-5.0.1-gsl2-branch
git tag -a XMI-MSIM-5.0.1 -m "5.0.1 maintenance tag for gsl2 compatibility"
git push origin XMI-MSIM-5.0.1
```

Then we can just build from the automatically generated https://github.com/tschoonj/xmimsim/archive/XMI-MSIM-5.0.1.tar.gz

>The main problem for me is that I cannot really write and test the patch until the gsl and fgsl formulas have both been updated to their latest versions and have been made available via brew update

Actually you can just do `brew pull 4341; brew upgrade -dvs gsl; brew upgrade -dvs fgsl` and then test as if they had already shipped. Be sure to `export HOMEBREW_DEVELOPER=1` and probably `export HOMEBREW_NO_AUTO_UPDATE=1` if you're working in the same tree as the brew command since we now auto update (which also stashes changes) whenever you run brew install or brew upgrade, and only pop the stash if HOMEBREW_DEVELOPER is set.

**Comment by tschoonj** [comment]
Thanks for long write-up!

I will get to work on it later today.

The maintenance tag thing won't work in my case since xmi-msim's binaries for Windows and Mac OS X that I distribute contain code that checks for updates by looking at the presence of new tags in the Github repo: if the corresponding binaries won't be available, then the user will be facing an error...

**Comment by ilovezfs** [comment]
>The maintenance tag thing won't work in my case since xmi-msim's binaries for Windows and Mac OS X that I distribute contain code that checks for updates by looking at the presence of new tags in the Github repo: if the corresponding binaries won't be available, then the user will be facing an error.

You can just temporarily fork the repo...

**Comment by tschoonj** [comment]
By the way: `fgsl` is in homebrew-science, not core. 

I suggest that I will take care of updating both xmi-msim and fgsl in the same PR: is that ok?

**Comment by ilovezfs** [comment]
For instance these two repos have different tags even though one is a fork of the other:

https://github.com/ouspg/radamsa/releases
https://github.com/aoh/radamsa/releases

But a patch file is also fine of course if you prefer to do it that way.

>By the way: fgsl is in homebrew-science, not core.

This was my only change for that:
```
diff --git a/fgsl.rb b/fgsl.rb
index c4fd6f1..ea50af6 100644
--- a/fgsl.rb
+++ b/fgsl.rb
@@ -1,9 +1,8 @@
 class Fgsl < Formula
   desc "Fortran bindings for the GNU Scientific Library"
   homepage "http://www.lrz.de/services/software/mathematik/gsl/fortran/"
-  url "http://www.lrz.de/services/software/mathematik/gsl/fortran/download/fgsl-1.0.0.tar.gz"
-  sha256 "2841f6deb2ce05e153fc1d89fe5e46aba74c60a2595c857cef9ca771a0cf6290"
-  revision 1
+  url "http://www.lrz.de/services/software/mathematik/gsl/fortran/download/fgsl-1.1.0.tar.gz"
+  sha256 "18d45e2bf87695587cd83320777ada0d9e2dfdee43eb9d76e0dab660758aff47"
 
   bottle do
     revision 1
```

>I suggest that I will take care of updating both xmi-msim and fgsl in the same PR: is that ok?

Yup, xmi-msim is the only reverse dependency of fgsl, so that works out nicely.

**Comment by ilovezfs** [comment]
OK so before this ships, the status quo in science is everything builds and tests fine except the following:

atomic-pseudopotential-engine: built, no test

hotspot: built, no test, fails audit
```
homebrew/science/hotspot:
  * Non-executables were installed to "/usr/local/Cellar/hotspot/4.1.0/bin"
    The offending files are:
      /usr/local/Cellar/hotspot/4.1.0/bin/wavelets
```

homebrew/science/libsequence: built, no test

homebrew/science/moose: build failed: https://gist.github.com/ilovezfs/1f699113e63da861ead7c8f8ebb97106

qgis: CI skipped it
```
==> SKIPPING: homebrew/science/qgis
Python could not import the PyQt4 module. This will cause the QGIS build to fail.
The most common reason for this failure is that the PYTHONPATH needs to be adjusted.
The pyqt caveats explain this adjustment and may be reviewed using:

```

File: tap_migrations.json [hunk 1]
```diff
@@ -195,6 +195,7 @@
   "openwsman": "homebrew/boneyard",
   "orfeo": "homebrew/boneyard",
   "ori": "homebrew/fuse",
+  "orpie": "homebrew/science",
```
Comment (ilovezfs) [inline tap_migrations.json]: should say boneyard. I fixed it in post.

**Comment by ilovezfs** [comment]
@tschoonj @sjackman @dpo shipped!

**Comment by ilovezfs** [comment]
PRs for the science revision bumps:

abinit: https://github.com/Homebrew/homebrew-science/pull/4177
analysis: https://github.com/Homebrew/homebrew-science/pull/4178
andi: https://github.com/Homebrew/homebrew-science/pull/4179
apophenia: https://github.com/Homebrew/homebrew-science/pull/4180
art: https://github.com/Homebrew/homebrew-science/pull/4181
astrometry-net: https://github.com/Homebrew/homebrew-science/pull/4182
atomic-pseudopotential-engine: https://github.com/Homebrew/homebrew-science/pull/4183
dynare: https://github.com/Homebrew/homebrew-science/pull/4184
ea-utils: https://github.com/Homebrew/homebrew-science/pull/4185
enblend-enfuse: https://github.com/Homebrew/homebrew-science/pull/4186
fwdpp: https://github.com/Homebrew/homebrew-science/pull/4187
getdp: https://github.com/Homebrew/homebrew-science/pull/4188
gnudatalanguage: https://github.com/Homebrew/homebrew-science/pull/4189
gromacs: https://github.com/Homebrew/homebrew-science/pull/4190
hotspot: https://github.com/Homebrew/homebrew-science/pull/4191
lobstr: https://github.com/Homebrew/homebrew-science/pull/4192
mash: https://github.com/Homebrew/homebrew-science/pull/4193
mathgl: https://github.com/Homebrew/homebrew-science/pull/4194
methpipe: https://github.com/Homebrew/homebrew-science/pull/4195
moose: https://github.com/Homebrew/homebrew-science/pull/4196
nco: https://github.com/Homebrew/homebrew-science/pull/4197
nest: https://github.com/Homebrew/homebrew-science/pull/4198
nip2: https://github.com/Homebrew/homebrew-science/pull/4199
pspp: https://github.com/Homebrew/homebrew-science/pull/4200
qgis: https://github.com/Homebrew/homebrew-science/pull/4201
root: https://github.com/Homebrew/homebrew-science/pull/4202
root6: https://github.com/Homebrew/homebrew-science/pull/4203
visp: https://github.com/Homebrew/homebrew-science/pull/4204

**Pull request closed without merging**